investigate(build):  warnings - gcc-12.2, ld-2.40

Starting with release 2.39 of the GNU Binutils several new warnings have been added to the linker.
Also gcc-12 may have new rules/checks.

In order to continue building nanos successfully ` -Wno-error=stringop-overflow -Wno-error=maybe-uninitialized` may need to be added to build flags. 

Posting all warnings, after adding those build flags, just to check if there is anything that's relevant and needs to be handled on code.

- **tools**
```
$ ld --version
GNU ld (GNU Binutils for Debian) 2.40
...
$ gcc --version
gcc (Debian 12.2.0-14) 12.2.0
```

- **build logs**
```
$ make                
...
ld: warning: /ops/nanos/output/klib/bin/firewall.dbg has a LOAD segment with RWX permissions
...
ld: warning: /ops/nanos/output/klib/bin/gcp.dbg has a LOAD segment with RWX permissions
...
ld: warning: /ops/nanos/output/klib/bin/ntp.dbg has a LOAD segment with RWX permissions
...
ld: warning: /ops/nanos/output/klib/bin/radar.dbg has a LOAD segment with RWX permissions
...
ld: warning: /ops/nanos/output/klib/bin/test/page_table.dbg has a LOAD segment with RWX permissions
...
ld: warning: /ops/nanos/output/platform/pc/boot/platform/pc/boot/service32.o: missing .note.GNU-stack section implies executable stack
ld: NOTE: This behaviour is deprecated and will be removed in a future version of the linker
ld: warning: /ops/nanos/output/platform/pc/boot/bin/stage2.elf has a LOAD segment with RWX permissions
...
ld: warning: /ops/nanos/output/platform/pc/boot/uefi/x86_64/uefi-crt0.o: missing .note.GNU-stack section implies executable stack
ld: NOTE: This behaviour is deprecated and will be removed in a future version of the linker
...
CC	/ops/nanos/output/platform/pc/src/unix/aio.o
In file included from /ops/nanos/src/runtime/runtime.h:3,
                 from /ops/nanos/src/kernel/kernel.h:2,
                 from /ops/nanos/src/unix/unix_internal.h:3,
                 from /ops/nanos/src/unix/aio.c:1:
In function ‘fetch_and_add’,
    inlined from ‘refcount_reserve’ at /ops/nanos/src/runtime/refcount.h:15:5,
    inlined from ‘aio_from_ring_id’ at /ops/nanos/src/unix/aio.c:68:5,
    inlined from ‘io_submit’ at /ops/nanos/src/unix/aio.c:303:11:
/ops/nanos/src/x86_64/machine.h:72:12: warning: ‘__sync_fetch_and_add_8’ writing 8 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
   72 |     return __sync_fetch_and_add(variable, value);
      |            ^~~~~~~~~~~~~~~~~~~~
/ops/nanos/src/x86_64/machine.h:72:12: warning: ‘__sync_fetch_and_add_8’ writing 8 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
At top level:
cc1: note: unrecognized command-line option ‘-Wno-unknown-warning-option’ may have been intended to silence earlier diagnostics
...
CC	/ops/nanos/output/platform/pc/src/unix/syscall.o
In file included from /ops/nanos/src/runtime/runtime.h:3,
                 from /ops/nanos/src/kernel/kernel.h:2,
                 from /ops/nanos/src/unix/unix_internal.h:3,
                 from /ops/nanos/src/unix/syscall.c:1:
In function ‘fetch_and_add’,
    inlined from ‘refcount_reserve’ at /ops/nanos/src/runtime/refcount.h:15:5,
    inlined from ‘thread_reserve’ at /ops/nanos/src/unix/unix_internal.h:617:5,
    inlined from ‘lookup_thread’ at /ops/nanos/src/unix/syscall.c:2268:9,
    inlined from ‘sched_setaffinity’ at /ops/nanos/src/unix/syscall.c:2281:15:
/ops/nanos/src/x86_64/machine.h:72:12: warning: ‘__sync_fetch_and_add_8’ writing 8 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
   72 |     return __sync_fetch_and_add(variable, value);
      |            ^~~~~~~~~~~~~~~~~~~~
In function ‘fetch_and_add’,
    inlined from ‘refcount_reserve’ at /ops/nanos/src/runtime/refcount.h:15:5,
    inlined from ‘thread_reserve’ at /ops/nanos/src/unix/unix_internal.h:617:5,
    inlined from ‘lookup_thread’ at /ops/nanos/src/unix/syscall.c:2268:9,
    inlined from ‘sched_getaffinity’ at /ops/nanos/src/unix/syscall.c:2300:15:
/ops/nanos/src/x86_64/machine.h:72:12: warning: ‘__sync_fetch_and_add_8’ writing 8 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
   72 |     return __sync_fetch_and_add(variable, value);
      |            ^~~~~~~~~~~~~~~~~~~~
At top level:
cc1: note: unrecognized command-line option ‘-Wno-unknown-warning-option’ may have been intended to silence earlier diagnostics
...
CC	/ops/nanos/output/platform/pc/src/hyperv/utilities/vmbus_ic.o
/ops/nanos/src/hyperv/utilities/vmbus_ic.c: In function ‘vmbus_ic_negomsg’:
/ops/nanos/src/hyperv/utilities/vmbus_ic.c:159:25: warning: ‘sel_fw_ver’ may be used uninitialized [-Wmaybe-uninitialized]
  159 |         nego->ic_ver[0] = sel_fw_ver;
      |         ~~~~~~~~~~~~~~~~^~~~~~~~~~~~
/ops/nanos/src/hyperv/utilities/vmbus_ic.c:46:18: note: ‘sel_fw_ver’ was declared here
   46 |         uint32_t sel_fw_ver, sel_msg_ver;
      |                  ^~~~~~~~~~
At top level:
cc1: note: unrecognized command-line option ‘-Wno-unknown-warning-option’ may have been intended to silence earlier diagnostics
...
ld: warning: /ops/nanos/output/platform/pc/src/x86_64/crt0.o: missing .note.GNU-stack section implies executable stack
ld: NOTE: This behaviour is deprecated and will be removed in a future version of the linker
...
```

--- a/platform/pc/boot/Makefile
+++ b/platform/pc/boot/Makefile
@@ -40,7 +40,7 @@ CFLAGS+= \
 	-I$(SRCDIR)/runtime \
 	-I$(SRCDIR)/kernel \
 	-I$(SRCDIR)/x86_64
-LDFLAGS+=	--gc-sections -T linker_script
+LDFLAGS+=	-z noexecstack --gc-sections -T linker_script
 AFLAGS+=	-I$(SRCDIR)/x86_64 -felf
 
 NASMFLAGS	= -l $@.lst -I $(SRCDIR)/x86_64/ -dSTAGE2SIZE=$(shell $(SIZE_CMD) $(OBJDIR)/stage2.pad)
@@ -97,7 +97,7 @@ CFLAGS-uefi= \
 	-I$(SRCDIR) -I$(SRCDIR)/boot -I$(SRCDIR)/kernel -I$(SRCDIR)/runtime \
 	-I$(SRCDIR)/fs -I$(SRCDIR)/x86_64 -I$(OBJDIR) \
 	-DBOOT -DUEFI -DEFIAPI="__attribute__((ms_abi))"
-LDLAGS-uefi= -nostdlib -shared -Bsymbolic -T $(SRCDIR)/x86_64/uefi.lds
+LDLAGS-uefi= -z noexecstack -nostdlib -shared -Bsymbolic -T $(SRCDIR)/x86_64/uefi.lds
 
 DEPFILES+= $(DEPS-uefi)
 

--- a/platform/pc/boot/linker_script
+++ b/platform/pc/boot/linker_script
@@ -3,12 +3,19 @@ OUTPUT_ARCH(i386)
 
 ENTRY(_start)
 
+PHDRS
+{
+  text PT_LOAD FLAGS(5);          /* R E */
+  rodata PT_LOAD FLAGS(4);        /* R */
+  data PT_LOAD FLAGS(6);          /* RW */
+}
+
 SECTIONS
 {
   . = 0x8000;
-  .start : { *(.start)}
-  .text : { *(.text)}
-  .rodata : { *(.rodata)}
-  .data : { *(.data) }
-  .bss  ALIGN(32): { *(.bss) }
+  .start : { *(.start)} :text
+  .text : { *(.text) *(.text.*) } :text
+  .rodata : { *(.rodata) *(.rodata.*) } :rodata
+  .data : { *(.data) *(.data.*) } :data
+  .bss ALIGN(32): { *(.bss) *(.bss.*) } :data
 }

--- a/platform/virt/Makefile
+++ b/platform/virt/Makefile
@@ -239,7 +239,7 @@ UEFI_CFLAGS=    -Wall -Werror -fpic -fshort-wchar -O3 -std=gnu11 \
 	-ffreestanding -DBOOT -DUEFI $(call cc-option, -mno-outline-atomics, "") \
 	-I$(SRCDIR) -I$(SRCDIR)/aarch64 -I$(SRCDIR)/boot -I$(SRCDIR)/kernel \
 	-I$(SRCDIR)/runtime -I$(SRCDIR)/fs -I$(PLATFORMDIR) -I$(OBJDIR)
-UEFI_LDFLAGS=   -nostdlib -shared -Bsymbolic -T $(SRCDIR)/aarch64/uefi.lds
+UEFI_LDFLAGS=   -z noexecstack -nostdlib -shared -Bsymbolic -T $(SRCDIR)/aarch64/uefi.lds
 
 DEPFILES+=      $(VDSO_DEPS) $(UEFI_DEPS)
 CLEANFILES+=    $(foreach f,gitversion.c frame.inc kernel.dis bin/kernel.elf boot/uefi.so src/unix/ftrace.* $(ARCHDIR)/ftrace.*,$(OBJDIR)/$f) $(VDSO_OBJDIR)/vdso.so $(VDSO_OBJDIR)/vdso-image.c $(VDSO_OBJDIR)/vdso-offset.h $(VDSO_OBJS) $(VDSO_DEPS) $(UEFI_OBJS) $(UEFI_DEPS) $(UEFI_LOADER) $(BOOTIMG) $(KERNEL)

--- a/rules.mk
+++ b/rules.mk
@@ -104,7 +104,7 @@ KERNCFLAGS+=	-march=rv64gc -mabi=lp64d
 endif
 
 KERNCFLAGS+=	-fno-omit-frame-pointer
-KERNLDFLAGS=	--gc-sections -z max-page-size=4096 -L $(OUTDIR)/klib -pie --no-dynamic-linker
+KERNLDFLAGS=	--gc-sections -z notext -z noexecstack -z max-page-size=4096 -L $(OUTDIR)/klib -pie --no-dynamic-linker
 
 ifneq ($(UBSAN),)
 KERNCFLAGS+= -fsanitize=undefined -fno-sanitize=alignment,null -fsanitize-undefined-trap-on-error

--- a/src/aarch64/klib.lds
+++ b/src/aarch64/klib.lds
@@ -2,14 +2,23 @@ OUTPUT_FORMAT("elf64-littleaarch64")
 
 ENTRY(init)
 
+PHDRS
+{
+        text PT_LOAD FLAGS(5);          /* R E */
+        rodata PT_LOAD FLAGS(4);        /* R */
+        data PT_LOAD FLAGS(6);          /* RW */
+}
+
 SECTIONS
 {
         . = SIZEOF_HEADERS;
-        .text : { *(.text)}
-        .rodata : { *(.rodata)}
+        .text : { *(.text) *(.text.*) } :text
+        . = ALIGN(4096);
+        .rodata : { *(.rodata) *(.rodata.*) } :rodata
 
-        .data ALIGN(4096) : { *(.data) *(.data.*) }
-        .bss ALIGN(32): { *(.bss) *(.bss.*) }
+        . = ALIGN(4096);
+        .data : { *(.data) *(.data.*) } :data
+        .bss : { *(.bss) *(.bss.*) } :data
 
         /DISCARD/ : { *(.interp) }
 }

--- a/src/riscv64/klib.lds
+++ b/src/riscv64/klib.lds
@@ -2,14 +2,23 @@ OUTPUT_FORMAT("elf64-littleriscv")
 
 ENTRY(init)
 
+PHDRS
+{
+        text PT_LOAD FLAGS(5);          /* R E */
+        rodata PT_LOAD FLAGS(4);        /* R */
+        data PT_LOAD FLAGS(6);          /* RW */
+}
+
 SECTIONS
 {
         . = SIZEOF_HEADERS;
-        .text : { *(.text)}
-        .rodata : { *(.rodata)}
+        .text : { *(.text) *(.text.*) } :text
+        . = ALIGN(4096);
+        .rodata : { *(.rodata) *(.rodata.*) } :rodata
 
-        .data ALIGN(4096) : { *(.data) *(.data.*) }
-        .bss ALIGN(32): { *(.bss) *(.bss.*) }
+        . = ALIGN(4096);
+        .data : { *(.data) *(.data.*) } :data
+        .bss : { *(.bss) *(.bss.*) } :data
 
         /DISCARD/ : { *(.interp) }
 }

--- a/src/x86_64/klib.lds
+++ b/src/x86_64/klib.lds
@@ -2,14 +2,23 @@ OUTPUT_FORMAT("elf64-x86-64")
 
 ENTRY(init)
 
+PHDRS
+{
+        text PT_LOAD FLAGS(5);          /* R E */
+        rodata PT_LOAD FLAGS(4);        /* R */
+        data PT_LOAD FLAGS(6);          /* RW */
+}
+
 SECTIONS
 {
         . = SIZEOF_HEADERS;
-        .text : { *(.text)}
-        .rodata : { *(.rodata)}
+        .text : { *(.text) *(.text.*) } :text
+        . = ALIGN(4096);
+        .rodata : { *(.rodata) *(.rodata.*) } :rodata
 
-        .data ALIGN(4096) : { *(.data) *(.data.*) }
-        .bss ALIGN(32): { *(.bss) *(.bss.*) }
+        . = ALIGN(4096);
+        .data : { *(.data) *(.data.*) } :data
+        .bss : { *(.bss) *(.bss.*) } :data
 
         /DISCARD/ : { *(.interp) }
 }

